hadoop-hdds/docs/content/design/storage-class.md

```diff
@@ -19,10 +19,331 @@ author: Marton Ele
+
+Therefore the STANDARD storage class can be modified to support three states: OPEN, WARM(=CLOSED), COLD. Transition between WARM and CLOSED can be managed by SCM (if container is not used, SCM can request to enable erasure coding). Or manually (one specific container can be forced to Erasure Coding / COLD state. Only useful for administrators)
+
+![Storage class transitions](storage-class-transition.png)
+
+## Storage class on the user interface
+
+Today, we need to define the replication factor (like `THREE`) and replication type (like `RATIS`) during the creation of the key. It's not possible to adjust possible replication rule for the key when it become part of a closed containers.
+
+With storage-class support the users don't need to udnerstand the details of the replication. Users can choose from the well-known or custom-defined storage-classes.
+
+# Implementation changes
+
+**Phase 1**: while storage class abstraction can provide a very generic framework for replication, in the first implementation It's proposed to keep it as simple as possible:
+
+ * Don't introduce any new custom container state
+ * Don't introduce any new transitions
+ * Don't introduce any new custom storage classes
+
+But we can start to use the abstraction for the existing implementation:
+
+ * We will implement two, hard-coded storage-classes (STANDARD, REDUCED) which can configure the behavior of existing replication schemes
+ * Ozone Client will be changed to support `--storage-class` instead of `--type` and `--factor`
+ * `ofs://` and `o3fs://` doesn't require any modification
+ * Storage class will be used as a custom String in RPC protocol (in storage-sensitive places it can be stored as an integer with an additional mapping)
+ * All the replication logic (PipelineManager/ReplicationManager) will work exactly as before. Storage-class will be resolved to the required replication config. Pipelines will have the same type as before (eg. Ratis/THREE)
+ * SCM allocateContainer logic should be changed. Right now it creates a predefined number of Open containers for each `owner` (high level application). It can be extended easily to create open containers for each storage-classes
+ * Having slightly more open containers doesn't make any difference (5-10 more open containers per disk vs. the 5000-10000 existing containers)
+
+# Backward compatibility
+
+ * In allocateBlock and intra-service communication we can replace replication type and replication factor with storage-class
+ * In client API replication factor and replication type will be optional
+ * Legacy type/factor definitions will be mapped to real storage classes under the hood (which defines the real replication scheme). For example `Ratis/THREE --> STANDARD`
+
```

Comment:
How will SCM replication monitor change?

Reply:
I wrote it down in more details. (Exactly as today but replication factor is coming from the storage-class configuration for each containers)

Reply:
It is not sufficient to just define the replication factor. The replication/recovery behavior completely changes based on the replication scheme. E.g. for EC, we are talking about picking some coordinator during recovery. How do you abstract away behavior like that?